python/ray/serve/_private/router.py

```diff
@@ -156,46 +108,70 @@ def update_autoscaling_config(self, autoscaling_config: AutoscalingConfig):
             # least one queued request on this router, then immediately
             # push handle metric to the controller.
-            if (
-                len(self._replica_scheduler.curr_replicas) == 0
-                and self.num_queued_queries
-            ):
-                self.push_metrics_to_controller(
-                    **self._get_aggregated_requests(), send_timestamp=time.time()
-                )
+            if self.should_send_scaled_to_zero_optimized_push(curr_num_replicas):
+                self.push_autoscaling_metrics_to_controller()
 
             if RAY_SERVE_COLLECT_AUTOSCALING_METRICS_ON_HANDLE:
                 # Record number of queued + ongoing requests at regular
                 # intervals into the in-memory metrics store
                 self.metrics_pusher.register_or_update_task(
-                    RECORD_METRICS_TASK_NAME,
+                    self.RECORD_METRICS_TASK_NAME,
                     self._add_autoscaling_metrics_point,
                     min(0.5, self.autoscaling_config.metrics_interval_s),
                 )
                 # Push metrics to the controller periodically.
                 self.metrics_pusher.register_or_update_task(
-                    PUSH_METRICS_TO_CONTROLLER_TASK_NAME,
-                    self._get_aggregated_requests,
+                    self.PUSH_METRICS_TO_CONTROLLER_TASK_NAME,
+                    self.push_autoscaling_metrics_to_controller,
                     self.autoscaling_config.metrics_interval_s,
-                    self.push_metrics_to_controller,
                 )
             else:
                 self.metrics_pusher.register_or_update_task(
-                    PUSH_METRICS_TO_CONTROLLER_TASK_NAME,
-                    self._get_aggregated_requests,
+                    self.PUSH_METRICS_TO_CONTROLLER_TASK_NAME,
+                    self.push_autoscaling_metrics_to_controller,
                     HANDLE_METRIC_PUSH_INTERVAL_S,
-                    self.push_metrics_to_controller,
                 )
 
             self.metrics_pusher.start()
         else:
             if self.metrics_pusher:
                 self.metrics_pusher.shutdown()
 
+    def inc_num_total_requests(self, route: str):
+        self.num_router_requests.inc(tags={"route": route})
+
+    def inc_num_queued_requests(self):
+        self.num_queued_requests += 1
+        self.num_queued_requests_gauge.set(self.num_queued_requests)
+
+    def dec_num_queued_requests(self):
+        self.num_queued_requests -= 1
+        self.num_queued_requests_gauge.set(self.num_queued_requests)
+
+    def inc_num_running_requests_for_replica(self, replica_tag: str):
+        self.num_requests_sent_to_replicas[replica_tag] += 1
+
+    def should_send_scaled_to_zero_optimized_push(self, curr_num_replicas: int) -> bool:
+        return (
+            self.autoscaling_config
+            and curr_num_replicas == 0
+            and self.num_queued_requests
```

Comment:
Yeah this would push the metrics multiple times if more requests arrive before a replica is started. I think this is probably fine since the record metrics task on the controller is pretty lightweight, and in case of bursty traffic, can probably help the deployment scale up faster. LMK if you have any concerns though.

Reply:
Yea, was thinking if burst event will kill the controller and maybe that's the reason it was checking just 1 before. But we will see🤞